### PR TITLE
store/tikv: support coprocessor streaming API for tikv client

### PR DESCRIPTION
**Pull request by tiancaiamao**
Implement the new gRPC coprocessor streaming API.
Some difficults emerge when we handle region split, or meet lock in the middle of the stream, retry process become complex. I refine some old code to adapt to that. 
When partial of the streaming result is locked, re-calculate the new range is error prone, those code need to be reviewed carefully.

### CODE REVIEW

**Comment by shenli** [comment]
Good job!

**Comment by tiancaiamao** [comment]
PTAL @disksing @coocood

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -456,84 +465,164 @@ func (it *copIterator) Next() ([]byte, error) {
 	return resp.Data, nil
 }
 
-// handleTask handles single copTask.
-func (it *copIterator) handleTask(bo *Backoffer, task *copTask) []copResponse {
-	coprocessorCounter.WithLabelValues("handle_task").Inc()
+// handleTasks handles single copTask, sends the result to channel, retry automatically on error.
+func (it *copIterator) handleTask(goCtx goctx.Context, bo *Backoffer, task *copTask, ch chan copResponse) {
+	var buf = [1]*copTask{task}
+	remainTasks := buf[:1]
```
Comment (disksing) [inline store/tikv/coprocessor.go]: why not `remainTasks := []*copTask{task}` directly?

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -456,84 +465,164 @@ func (it *copIterator) Next() ([]byte, error) {
...
+	if desc {
+		i := len - 1
+		for ; i >= 0; i-- {
+			r := ranges.at(i)
+			if bytes.Compare(r.StartKey, split.End) < 0 {
+				break
+			}
+		}
+		if i < 0 {
+			log.Error("wrong response from tikv!")
```
Comment (disksing) [inline store/tikv/coprocessor.go]: log more details or return an error.

File: store/tikv/coprocessor.go [hunk 3]
```diff
@@ -456,84 +465,164 @@ func (it *copIterator) Next() ([]byte, error) {
...
-		return []copResponse{{err: errors.Trace(err)}}
+	ranges := calculateRemain(task.ranges, resp.Range, desc)
+	return buildCopTasks(bo, cache, ranges, desc, streaming)
+}
+
+func calculateRemain(ranges *copRanges, split *coprocessor.KeyRange, desc bool) *copRanges {
+	len := ranges.len()
+	if desc {
+		i := len - 1
+		for ; i >= 0; i-- {
```
Comment (disksing) [inline store/tikv/coprocessor.go]: I think we need binary search for desc too.

File: store/tikv/coprocessor.go (hunk 1 above)
Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: An optimization to make less object allocation, at the cost of readability, I'll change here.

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: I think it's feasible to panic here.
If TiKV's behavior is right, the code should never run to here.

**Comment by disksing** [comment]
LGTM. Although I think `calculateRemain` can be more simpler :)

**Comment by tiancaiamao** [comment]
PTAL @coocood @zz-jason

File: store/tikv/coprocessor.go [hunk 4]
```diff
@@ -456,84 +465,158 @@ func (it *copIterator) Next() ([]byte, error) {
...
+func (it *copIterator) handleCopStreamResult(goCtx goctx.Context, bo *Backoffer, stream tikvpb.Tikv_CoprocessorStreamClient, task *copTask, ch chan copResponse) ([]*copTask, error) {
 	for {
-		select {
-		case <-it.finished:
-			return nil
-		default:
+		resp, err := stream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
```
Comment (zz-jason) [inline store/tikv/coprocessor.go]: `return nil, nil` can be more direct.

File: store/tikv/coprocessor.go [hunk 5]
```diff
@@ -456,84 +465,158 @@ func (it *copIterator) Next() ([]byte, error) {
...
-				Tp:     it.req.Tp,
-				Data:   it.req.Data,
-				Ranges: task.ranges.toPBRanges(),
-			},
-			Context: kvrpcpb.Context{
-				IsolationLevel: pbIsolationLevel(it.req.IsolationLevel),
-				Priority:       kvPriorityToCommandPri(it.req.Priority),
-				NotFillCache:   it.req.NotFillCache,
-			},
+		tasks, err := it.checkCopResponse(bo, resp, task)
```
Comment (zz-jason) [inline store/tikv/coprocessor.go]: s/tasks/remainedTasks/

File: store/tikv/coprocessor.go [hunk 6]
```diff
@@ -456,84 +465,158 @@ func (it *copIterator) Next() ([]byte, error) {
...
 
-	ret := make([]copResponse, 0, len(newTasks))
-	for _, t := range newTasks {
-		resp := it.handleTask(bo, t)
-		ret = append(ret, resp...)
+	ranges := calculateRemain(task.ranges, resp.Range, desc)
+	return buildCopTasks(bo, cache, ranges, desc, streaming)
+}
+
+func calculateRemain(ranges *copRanges, split *coprocessor.KeyRange, desc bool) *copRanges {
```
Comment (zz-jason) [inline store/tikv/coprocessor.go]: s/ranges/originRanges/
s/split/failedRange/

File: store/tikv/coprocessor.go [hunk 7]
```diff
@@ -456,84 +465,158 @@ func (it *copIterator) Next() ([]byte, error) {
...
+		n := sort.Search(len, func(i int) bool {
+			r := ranges.at(i)
+			return bytes.Compare(split.End, r.StartKey) <= 0
+		})
+		if n == 0 {
+			panic("for any valid tikv response, n > 0 should always hold!")
+		}
+		r := ranges.at(n - 1)
+		ret := ranges.slice(0, n-1)
+		ret.last = &kv.KeyRange{r.StartKey, r.EndKey}
```
Comment (zz-jason) [inline store/tikv/coprocessor.go]: add comment: `// [r.StartKey -- split.EndKey -- r.EndKey) => [r.StartKey -- split.EndKey)`

File: store/tikv/coprocessor.go [hunk 8]
```diff
@@ -408,6 +409,14 @@ func (it *copIterator) sendToTaskCh(ctx goctx.Context, t *copTask, taskCh chan<-
 	return
 }
 
+func (it *copIterator) sendToRespCh(goCtx goctx.Context, resp copResponse, respCh chan copResponse) {
+	select {
+	case respCh <- resp:
+	case <-it.finished:
```
Comment (zz-jason) [inline store/tikv/coprocessor.go]: we should add a mechanism to make the current goroutine exit when `it.finished` is closed

Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: Nice catch! I'll fix here.

File: store/tikv/coprocessor.go (hunk 6 above)
Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: This change is unnecessary, `ranges` and `split` is straightforward enough.

File: store/tikv/coprocessor.go [hunk 9]
```diff
@@ -456,84 +465,158 @@ func (it *copIterator) Next() ([]byte, error) {
...
-	coprocessorCounter.WithLabelValues("rebuild_task").Inc()
-
-	newTasks, err := buildCopTasks(bo, it.store.regionCache, task.ranges, it.req.Desc)
-	if err != nil {
-		return []copResponse{{err: errors.Trace(err)}}
-	}
-	if newTasks == nil {
-		// TODO: check this, this should never happen.
-		return nil
+func buildCopTasksFromRemain(bo *Backoffer, cache *RegionCache, resp *coprocessor.Response, task *copTask, desc bool, streaming bool) ([]*copTask, error) {
```
Comment (coocood) [inline store/tikv/coprocessor.go]: How about
```
remainedRanges := task.ranges
if streaming {
    remainedRanges = calculateRemain(task.ranges, resp.Range, desc)
}
return buildCopTasks(bo, cache, ranges, desc, streaming)
```

And I think inline this function is easier to read.

File: store/tikv/coprocessor.go (hunk 6 above)
Comment (coocood) [reply store/tikv/coprocessor.go]: or `retryRange`

File: store/tikv/coprocessor.go [hunk 10]
```diff
@@ -456,84 +465,158 @@ func (it *copIterator) Next() ([]byte, error) {
...
-		if err != nil {
-			return []copResponse{{err: errors.Trace(err)}}
+
+		it.sendToRespCh(goCtx, copResponse{resp, nil}, ch)
+	}
+	return nil, nil
+}
+
+// checkCopResponse checks coprocessor Response for region split and lock.
+func (it *copIterator) checkCopResponse(bo *Backoffer, resp *coprocessor.Response, task *copTask) ([]*copTask, error) {
```
Comment (coocood) [inline store/tikv/coprocessor.go]: The function name doesn't describe its behavior.
I think it's better to only check the response, and return a bool to indicate if the task needs rebuild, then explicitly rebuild the task outside of the function.

File: store/tikv/coprocessor.go (hunk 6 above)
Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: I think it's good enough now. `retryRange` introduce ambiguity: the returned result may also called `retryRange`

File: store/tikv/coprocessor.go [hunk 11]
```diff
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
...
+		// [r.StartKey -- split.EndKey -- r.EndKey) => [r.StartKey -- split.EndKey)
+		if bytes.Compare(split.End, r.EndKey) < 0 {
+			ret.last.EndKey = split.End
+		}
+		return ret
 	}
-	return ret
+
+	key := split.Start
+	// Binary search for the range r that [r, r+1) is not consumed.
```
Comment (coocood) [inline store/tikv/coprocessor.go]: What does `[r, r+1)` mean? I think it's easier to under stand without it.

File: store/tikv/coprocessor.go [hunk 12]
```diff
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
...
+	n := sort.Search(len, func(i int) bool {
+		r := ranges.at(i)
+		return bytes.Compare(r.EndKey, key) > 0
+	})
+	if n >= len {
+		log.Fatal("for any valid tikv response, n < len should always hold!")
```
Comment (coocood) [inline store/tikv/coprocessor.go]: Do not `log.Fatal`
just let it panic if n >= len.

File: store/tikv/coprocessor.go [hunk 13]
```diff
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
...
-		ret = append(ret, resp...)
+func calculateRemain(ranges *copRanges, split *coprocessor.KeyRange, desc bool) *copRanges {
+	len := ranges.len()
+	if desc {
+		n := sort.Search(len, func(i int) bool {
+			r := ranges.at(i)
+			return bytes.Compare(split.End, r.StartKey) <= 0
+		})
+		if n == 0 {
+			log.Fatal("for any valid tikv response, n > 0 should always hold!")
```
Comment (coocood) [inline store/tikv/coprocessor.go]: Do not `log.Fatal` it exits the process instead of the session.

File: store/tikv/coprocessor.go [hunk 14]
```diff
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
...
-	ret := make([]copResponse, 0, len(newTasks))
-	for _, t := range newTasks {
-		resp := it.handleTask(bo, t)
-		ret = append(ret, resp...)
+func calculateRemain(ranges *copRanges, split *coprocessor.KeyRange, desc bool) *copRanges {
+	len := ranges.len()
+	if desc {
+		n := sort.Search(len, func(i int) bool {
+			r := ranges.at(i)
+			return bytes.Compare(split.End, r.StartKey) <= 0
```
Comment (coocood) [inline store/tikv/coprocessor.go]: Why not `bytes.Compare(r.StartKey, split.End) > 0`?

File: store/tikv/coprocessor.go [hunk 15]
```diff
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
...
+}
 
-	ret := make([]copResponse, 0, len(newTasks))
-	for _, t := range newTasks {
-		resp := it.handleTask(bo, t)
-		ret = append(ret, resp...)
+func calculateRemain(ranges *copRanges, split *coprocessor.KeyRange, desc bool) *copRanges {
+	len := ranges.len()
+	if desc {
+		n := sort.Search(len, func(i int) bool {
```
Comment (coocood) [inline store/tikv/coprocessor.go]: The name `n` is meaningless, how about `minConsumedIdx`?

File: store/tikv/coprocessor.go [hunk 16]
```diff
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
...
+		if bytes.Compare(split.End, r.EndKey) < 0 {
+			ret.last.EndKey = split.End
+		}
+		return ret
 	}
-	return ret
+
+	key := split.Start
+	// Binary search for the range r that [r, r+1) is not consumed.
+	n := sort.Search(len, func(i int) bool {
```
Comment (coocood) [inline store/tikv/coprocessor.go]: How about `minNotConsumedIdx`?

File: store/tikv/coprocessor.go [hunk 17]
```diff
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
...
+		r := ranges.at(i)
+		return bytes.Compare(r.EndKey, key) > 0
+	})
+	if n >= len {
+		log.Fatal("for any valid tikv response, n < len should always hold!")
+	}
+	r := ranges.at(n)
```
Comment (coocood) [inline store/tikv/coprocessor.go]: minNotConsumedRange

File: store/tikv/coprocessor.go [hunk 18]
```diff
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
...
+	len := ranges.len()
+	if desc {
+		n := sort.Search(len, func(i int) bool {
+			r := ranges.at(i)
+			return bytes.Compare(split.End, r.StartKey) <= 0
+		})
+		if n == 0 {
+			log.Fatal("for any valid tikv response, n > 0 should always hold!")
+		}
+		r := ranges.at(n - 1)
```
Comment (coocood) [inline store/tikv/coprocessor.go]: minNotConsumedRange

**Comment by disksing** [comment]
I think we can add a split method to `copRanges`:
```
// Split ranges into (left, right) by key.
func (r *copRanges) split(key []byte) (*copRanges, *copRanges) {
	n := sort.Search(r.len(), func(i int) bool {
		cur := r.at(i)
		return len(cur.EndKey) == 0 || bytes.Compare(cur.EndKey, key) > 0
	})
	// If a range p contains the key, it will split to 2 parts.
	if n < r.len() {
		p := r.at(n)
		if bytes.Compare(key, p.StartKey) > 0 {
			left := r.slice(0, n)
			left.last = &kv.KeyRange{StartKey: p.StartKey, EndKey: key}
			right := r.slice(n+1, r.len())
			right.first = &kv.KeyRange{StartKey: key, EndKey: p.EndKey}
			return left, right
		}
	}
	return r.slice(0, n), r.slice(n, r.len())
}
```
Then `calculateRemain` can be very simple:
```
func calculateRemain(ranges *copRanges, split *coprocessor.KeyRange, desc bool) *copRanges {
	if desc {
		left, _ := ranges.split(split.End)
		return left
	}
	_, right := ranges.split(split.Start)
	return right
}
```

File: store/tikv/coprocessor.go (hunk 13 above)
Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: 1. It's called in a worker goroutine, so this line will not make process exit.
2. it should never happen.

File: store/tikv/coprocessor.go (hunk 9 above)
Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: Done

File: store/tikv/coprocessor_test.go [hunk 19]
```diff
@@ -162,3 +163,181 @@ func (s *testCoprocessorSuite) checkEqual(c *C, copRanges *copRanges, ranges []k
...
+}
+
+func coprocessorKeyRange(start, end string) *coprocessor.KeyRange {
+	return &coprocessor.KeyRange{
+		Start: []byte(start),
+		End:   []byte(end),
+	}
+}
+
+type zzzCase struct {
```
Comment (coocood) [inline store/tikv/coprocessor_test.go]: What does `zzz` means?

Comment (tiancaiamao) [reply store/tikv/coprocessor_test.go]: nothing...any suggestion?

Comment (coocood) [reply store/tikv/coprocessor_test.go]: Since we added `split` method, then we can just test `split`.

Comment (tiancaiamao) [reply store/tikv/coprocessor_test.go]: Good idea. I'll change it.

Comment (tiancaiamao) [reply store/tikv/coprocessor_test.go]: Done.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
/run-all-tests

@disksing @zz-jason

**Pull request merged**